File: src/telemetry.ts
import { monitorEventLoopDelay } from "node:perf_hooks";
import { performance } from "node:perf_hooks";
import { OTLPMetricExporter } from "@opentelemetry/exporter-metrics-otlp-http";
import { resourceFromAttributes } from "@opentelemetry/resources";
import { MeterProvider, PeriodicExportingMetricReader } from "@opentelemetry/sdk-metrics";
import type { ApolloServerPlugin } from "@apollo/server";
import { Kind } from "graphql";
import type { DocumentNode, OperationDefinitionNode, SelectionSetNode } from "graphql";
import type { PrismaClient } from "@prisma/client";

// Pushes metrics as OTLP to the same OTel Collector the old Micrometer/OTLP registry
// used (OTEL_METRICS_URL, unchanged in k8s/configmap.yaml), tagged with the same
// k8s.pod.name resource attribute (POD_NAME, unchanged Downward API env in
// k8s/deployment.yaml) so per-pod dashboards keep working exactly as before.
const OTEL_METRICS_URL = process.env.OTEL_METRICS_URL ?? "http://localhost:4318/v1/metrics";
const POD_NAME = process.env.POD_NAME ?? "local";

const resource = resourceFromAttributes({
  "service.name": "message-service",
  "k8s.pod.name": POD_NAME,
});

const exporter = new OTLPMetricExporter({ url: OTEL_METRICS_URL });

const meterProvider = new MeterProvider({
  resource,
  readers: [new PeriodicExportingMetricReader({ exporter, exportIntervalMillis: 15000 })],
});

const meter = meterProvider.getMeter("message-service");

// No `unit` option on "_ms"-suffixed instruments below: the OTel Collector's Prometheus
// exporter appends its own unit-derived suffix ("ms" -> "_milliseconds") whenever a name
// doesn't already end in that exact translated string, which double-suffixes a name that
// already spells out "_ms" (confirmed live: `graphql_request_duration_ms_milliseconds_bucket`).
// Names already carry their unit, so unit hints are intentionally left off everywhere.

// Replaces http.server.requests (Micrometer) - labeled by the schema's own operation type
// and root field (see metricsPlugin) instead of HTTP method+route, since every request here
// is POST /graphql. Deliberately NOT labeled by the client-supplied operationName: it's
// unbounded (any client can invent names), which would let callers grow the series count
// in the collector and in OpenObserve at will.
const graphqlRequestCounter = meter.createCounter("graphql_requests_total", {
  description:
    "Count of GraphQL operations executed, incremented once per root field of the operation",
});
const graphqlErrorCounter = meter.createCounter("graphql_errors_total", {
  description:
    "Count of GraphQL operations that returned errors, once per root field and distinct error code",
});
const graphqlRequestDuration = meter.createHistogram("graphql_request_duration_ms", {
  description:
    "GraphQL operation duration in milliseconds; a multi-root-field operation records its full duration once per root field",
});

// Replaces JVM thread/GC pressure signals - there's no JVM anymore, so event-loop lag
// is the equivalent "is the runtime falling behind" saturation signal for Node.
const eventLoopDelay = monitorEventLoopDelay({ resolution: 20 });
eventLoopDelay.enable();
meter
  .createObservableGauge("nodejs_eventloop_lag_p99_ms", {
    description: "Node.js event loop delay, 99th percentile, milliseconds",
  })
  .addCallback((result) => {
    result.observe(eventLoopDelay.percentile(99) / 1e6);
  });

meter
  .createObservableGauge("nodejs_process_memory_rss_bytes", { description: "Resident set size" })
  .addCallback((result) => result.observe(process.memoryUsage().rss));

meter
  .createObservableGauge("nodejs_process_memory_heap_used_bytes", { description: "V8 heap used" })
  .addCallback((result) => result.observe(process.memoryUsage().heapUsed));

let lastCpuUsage = process.cpuUsage();
let lastCpuSampleAt = process.hrtime.bigint();
meter
  .createObservableGauge("nodejs_process_cpu_usage_ratio", {
    description: "Process CPU usage as a fraction of one core since the last sample",
  })
  .addCallback((result) => {
    const usage = process.cpuUsage(lastCpuUsage);
    const now = process.hrtime.bigint();
    const elapsedNs = Number(now - lastCpuSampleAt);
    const cpuNs = (usage.user + usage.system) * 1000;
    result.observe(elapsedNs > 0 ? cpuNs / elapsedNs : 0);
    lastCpuUsage = process.cpuUsage();
    lastCpuSampleAt = now;
  });

// Direct replacement for the old HikariCP pool metrics: Prisma's own connection-pool
// and query-timing stats (prisma.$metrics.json(), needs the "metrics" preview feature
// in prisma/schema.prisma), sampled once per collection tick via a batch callback
// rather than one client.$metrics.json() round trip per gauge.
export function registerPrismaMetrics(prisma: PrismaClient): void {
  const poolOpen = meter.createObservableGauge("prisma_pool_connections_open");
  const poolBusy = meter.createObservableGauge("prisma_pool_connections_busy");
  const poolIdle = meter.createObservableGauge("prisma_pool_connections_idle");
  const queryDurationAvg = meter.createObservableGauge("prisma_client_queries_duration_avg_ms");
  const queryWaitAvg = meter.createObservableGauge("prisma_client_queries_wait_avg_ms");

  meter.addBatchObservableCallback(
    async (result) => {
      const metrics = await prisma.$metrics.json();
      const gauge = (key: string): number =>
        metrics.gauges.find((entry) => entry.key === key)?.value ?? 0;
      const histogramAvg = (key: string): number => {
        const histogram = metrics.histograms.find((entry) => entry.key === key)?.value;
        if (!histogram || histogram.count === 0) return 0;
        return histogram.sum / histogram.count;
      };

      result.observe(poolOpen, gauge("prisma_pool_connections_open"));
      result.observe(poolBusy, gauge("prisma_pool_connections_busy"));
      result.observe(poolIdle, gauge("prisma_pool_connections_idle"));
      result.observe(queryDurationAvg, histogramAvg("prisma_client_queries_duration_histogram_ms"));
      result.observe(queryWaitAvg, histogramAvg("prisma_client_queries_wait_histogram_ms"));
    },
    [poolOpen, poolBusy, poolIdle, queryDurationAvg, queryWaitAvg],
  );
}

// Root fields of the operation, as response key (the alias if there is one) -> schema field
// name, looking through inline fragments and fragment spreads. Only called from
// didResolveOperation, i.e. after validation: every field name is a real schema field, so
// the root_field label's cardinality is bounded by the schema, and fragment cycles are
// already ruled out. The response key is kept so a field error (whose path starts with it)
// can be attributed to the root field that actually failed.
function rootFieldsByResponseKey(
  document: DocumentNode,
  operation: OperationDefinitionNode,
): Map<string, string> {
  const fragments = new Map<string, SelectionSetNode>();
  for (const definition of document.definitions) {
    if (definition.kind === Kind.FRAGMENT_DEFINITION) {
      fragments.set(definition.name.value, definition.selectionSet);
    }
  }
  const fields = new Map<string, string>();
  const collect = (selectionSet: SelectionSetNode): void => {
    for (const selection of selectionSet.selections) {
      if (selection.kind === Kind.FIELD) {
        fields.set(selection.alias?.value ?? selection.name.value, selection.name.value);
      } else if (selection.kind === Kind.INLINE_FRAGMENT) {
        collect(selection.selectionSet);
      } else {
        const fragment = fragments.get(selection.name.value);
        if (fragment) collect(fragment);
      }
    }
  };
  collect(operation.selectionSet);
  return fields;
}

// Requests that fail before an operation is resolved (parse / validation errors) have no
// operation type or root field - they're bucketed under this placeholder and told apart by
// error_code (GRAPHQL_PARSE_FAILED / GRAPHQL_VALIDATION_FAILED).
const UNRESOLVED = "unresolved";

export const metricsPlugin: ApolloServerPlugin = {
  async requestDidStart() {
    const start = performance.now();
    let operationType = UNRESOLVED;
    let fieldByResponseKey = new Map<string, string>();
    let rootFields = [UNRESOLVED];
    return {
      async didResolveOperation({ document, operation }) {
        if (!operation) return;
        operationType = operation.operation;
        fieldByResponseKey = rootFieldsByResponseKey(document, operation);
        if (fieldByResponseKey.size > 0) rootFields = [...new Set(fieldByResponseKey.values())];
      },
      async willSendResponse({ response }) {
        const durationMs = performance.now() - start;

        // Attribute each error to the root field it came from (its path starts with that
        // field's response key). An error with no such path - e.g. a validation or
        // variable-coercion failure - can't be pinned to one field, so it counts against
        // every root field of the operation. extensions.code is only ever set by this
        // service's own errors.ts helpers or by Apollo itself, so it's a small closed set.
        const errorCodesByField = new Map<string, Set<string>>();
        if (response.body.kind === "single") {
          for (const error of response.body.singleResult.errors ?? []) {
            const code = String(error.extensions?.code ?? "UNKNOWN");
            const responseKey = error.path?.[0];
            const field =
              typeof responseKey === "string" ? fieldByResponseKey.get(responseKey) : undefined;
            for (const affected of field ? [field] : rootFields) {
              const codes = errorCodesByField.get(affected) ?? new Set<string>();
              errorCodesByField.set(affected, codes.add(code));
            }
          }
        }

        for (const rootField of rootFields) {
          const attributes = { operation_type: operationType, root_field: rootField };
          graphqlRequestDuration.record(durationMs, attributes);
          graphqlRequestCounter.add(1, attributes);
          for (const errorCode of errorCodesByField.get(rootField) ?? []) {
            graphqlErrorCounter.add(1, { ...attributes, error_code: errorCode });
          }
        }
      },
    };
  },
};

export async function shutdownTelemetry(): Promise<void> {
  await meterProvider.shutdown();
}

File: app/telemetry.py
"""Metrics: OTLP/HTTP push to the collector every 15s (there is no /metrics scrape endpoint).

Resource attributes service.name and k8s.pod.name become the Prometheus labels service_name and
k8s_pod_name (the collector has resource_to_telemetry_conversion on) - so they are NOT repeated as
data-point attributes. No `unit` is set on any instrument: the collector's Prometheus exporter
appends a unit-derived suffix ("ms" -> "_milliseconds") unless the name already ends in exactly
that, which would double-suffix names that already spell out `_ms` / `_bytes`.
"""

import asyncio
import logging
import time
from collections import deque
from collections.abc import Callable, Iterable
from typing import Any

import psutil
from opentelemetry import metrics
from opentelemetry.exporter.otlp.proto.http.metric_exporter import OTLPMetricExporter
from opentelemetry.metrics import CallbackOptions, Observation
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import Resource
from sqlalchemy import event
from sqlalchemy.ext.asyncio import AsyncEngine
from starlette.types import ASGIApp, Receive, Scope, Send

from app.errors import problem_response
from app.settings import get_settings

log = logging.getLogger("app.access")

SERVICE_NAME = "message-service"
# Bounds every export attempt (and so shutdown): an unreachable collector must not hold the pod
# past terminationGracePeriodSeconds.
EXPORT_TIMEOUT_S = 5

# Instruments come from the global meter: no-ops until setup_metrics() installs a provider, so
# the middleware works unchanged in tests.
_meter = metrics.get_meter(SERVICE_NAME)
_requests = _meter.create_counter(
    "http_requests_total", description="HTTP requests handled, by method, route template and status"
)
_errors = _meter.create_counter(
    "http_errors_total", description="HTTP error responses, by method, route template and code"
)
_duration = _meter.create_histogram(
    "http_request_duration_ms", description="HTTP request duration in milliseconds"
)

# Kubelet probes hit these every few seconds; they would drown the real traffic.
EXCLUDED_PREFIX = "/health/"
UNMATCHED_ROUTE = "unmatched"


class RequestMetricsMiddleware:
    """Pure ASGI middleware (no BaseHTTPMiddleware: it breaks contextvars and streaming).

    Also the last line of defence for unhandled exceptions: logs them with the active trace id and
    answers a generic problem+json 500, so no stack trace ever reaches the client.
    """

    def __init__(self, app: ASGIApp) -> None:
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if scope["type"] != "http" or scope["path"].startswith(EXCLUDED_PREFIX):
            await self.app(scope, receive, send)
            return

        start = time.perf_counter()
        status = 500
        started = False

        async def send_wrapper(message: Any) -> None:
            nonlocal status, started
            if message["type"] == "http.response.start":
                status = message["status"]
                started = True
            await send(message)

        try:
            await self.app(scope, receive, send_wrapper)
        except Exception:
            log.exception("unhandled exception", extra={"method": scope["method"]})
            status = 500
            if not started:
                response = problem_response(
                    scope, 500, "INTERNAL_SERVER_ERROR", "An unexpected error occurred."
                )
                await response(scope, receive, send_wrapper)
        finally:
            elapsed_ms = (time.perf_counter() - start) * 1000
            route = getattr(scope.get("route"), "path", UNMATCHED_ROUTE)
            attrs = {"method": scope["method"], "route": route}
            _duration.record(elapsed_ms, attrs)
            _requests.add(1, {**attrs, "status_code": status})
            if status >= 400:
                _errors.add(1, {**attrs, "error_code": scope.get("error_code", "UNKNOWN")})
            log.info(
                "request",
                extra={
                    "method": scope["method"],
                    "route": route,
                    "status_code": status,
                    "duration_ms": round(elapsed_ms, 2),
                },
            )


class _LoopLagMonitor:
    """Samples how late a 20ms sleep wakes up: the event-loop saturation signal for asyncio."""

    INTERVAL = 0.02

    def __init__(self) -> None:
        self._samples: deque[float] = deque(maxlen=3000)  # ~60s of history
        self._task: asyncio.Task[None] | None = None

    def start(self) -> None:
        self._task = asyncio.get_running_loop().create_task(self._run())

    async def stop(self) -> None:
        if self._task:
            self._task.cancel()
            await asyncio.gather(self._task, return_exceptions=True)

    async def _run(self) -> None:
        while True:
            before = time.perf_counter()
            await asyncio.sleep(self.INTERVAL)
            self._samples.append(max((time.perf_counter() - before - self.INTERVAL) * 1000, 0.0))

    def p99(self) -> float:
        if not self._samples:
            return 0.0
        ordered = sorted(self._samples)
        return ordered[int(0.99 * (len(ordered) - 1))]


class _QueryStats:
    """Average SQL statement duration since the previous collection."""

    def __init__(self) -> None:
        self._total_ms = 0.0
        self._count = 0

    def add(self, ms: float) -> None:
        self._total_ms += ms
        self._count += 1

    def take_average(self) -> float:
        avg = self._total_ms / self._count if self._count else 0.0
        self._total_ms, self._count = 0.0, 0
        return avg


_provider: MeterProvider | None = None
_lag = _LoopLagMonitor()


def setup_metrics(engine: AsyncEngine) -> None:
    global _provider
    settings = get_settings()
    # Resource() rather than Resource.create(): no telemetry.sdk.* attributes, which the
    # collector would turn into extra labels on every series.
    resource = Resource({"service.name": SERVICE_NAME, "k8s.pod.name": settings.pod_name})
    reader = PeriodicExportingMetricReader(
        OTLPMetricExporter(endpoint=settings.otel_metrics_url, timeout=EXPORT_TIMEOUT_S),
        export_interval_millis=15000,
    )
    _provider = MeterProvider(resource=resource, metric_readers=[reader])
    metrics.set_meter_provider(_provider)
    meter = _provider.get_meter(SERVICE_NAME)

    def gauge(name: str, description: str, read: Callable[[], float]) -> None:
        def callback(_: CallbackOptions) -> Iterable[Observation]:
            yield Observation(read())

        meter.create_observable_gauge(name, callbacks=[callback], description=description)

    process = psutil.Process()
    gauge(
        "python_process_memory_rss_bytes",
        "Resident set size",
        lambda: float(process.memory_info().rss),
    )

    last = {"cpu": sum(process.cpu_times()[:2]), "wall": time.monotonic()}

    def cpu_ratio() -> float:
        cpu, wall = sum(process.cpu_times()[:2]), time.monotonic()
        elapsed = wall - last["wall"]
        ratio = (cpu - last["cpu"]) / elapsed if elapsed > 0 else 0.0
        last["cpu"], last["wall"] = cpu, wall
        return ratio

    gauge(
        "python_process_cpu_usage_ratio",
        "Process CPU usage as a fraction of one core since the last sample",
        cpu_ratio,
    )

    _lag.start()
    gauge(
        "python_eventloop_lag_p99_ms",
        "asyncio event loop scheduling delay, 99th percentile over ~60s, milliseconds",
        _lag.p99,
    )

    # DB pool + query timing, read from the SQLAlchemy pool and cursor-execute events.
    pool: Any = engine.sync_engine.pool
    gauge(
        "db_pool_connections_open",
        "Connections currently open (busy + idle)",
        lambda: float(pool.checkedin() + pool.checkedout()),
    )
    gauge("db_pool_connections_busy", "Connections checked out", lambda: float(pool.checkedout()))
    gauge(
        "db_pool_connections_idle", "Connections idle in the pool", lambda: float(pool.checkedin())
    )

    stats = _QueryStats()

    @event.listens_for(engine.sync_engine, "before_cursor_execute")
    def _before(
        conn: Any, cursor: Any, statement: Any, parameters: Any, context: Any, many: Any
    ) -> None:
        context._query_started = time.perf_counter()

    @event.listens_for(engine.sync_engine, "after_cursor_execute")
    def _after(
        conn: Any, cursor: Any, statement: Any, parameters: Any, context: Any, many: Any
    ) -> None:
        stats.add((time.perf_counter() - context._query_started) * 1000)

    gauge(
        "db_client_queries_duration_avg_ms",
        "Average SQL statement duration since the previous collection, milliseconds",
        stats.take_average,
    )


async def shutdown_metrics() -> None:
    await _lag.stop()
    if _provider is not None:
        # shutdown() performs a final collect + export, so no separate force_flush is needed.
        await asyncio.to_thread(_provider.shutdown, EXPORT_TIMEOUT_S * 1000)
